// yass/yass.ts
export class Writer {
    private capacity: number;
    private position = 0;
    array: Uint8Array;
    constructor(initialCapacity: number) {
        this.capacity = initialCapacity;
        this.array = new Uint8Array(initialCapacity);
    }
    needed(value: number): number {
        const oldPosition = this.position;
        this.position += value;
        if (this.position > this.capacity) {
            const oldArray = this.array;
            this.capacity = 2 * this.position;
            this.array = new Uint8Array(this.capacity);
            this.array.set(oldArray);
        }
        return oldPosition;
    }
    writeByte(value: number): void {
        const position = this.needed(1);
        this.array[position] = value;
    }
    writeInt(value: number): void {
        const position = this.needed(4);
        new DataView(this.array.buffer).setInt32(position, value);
    }
    writeVarInt(value: number): void {
        while (true) {
            if ((value & ~0x7F) === 0) {
                this.writeByte(value);
                return;
            }
            this.writeByte((value & 0x7F) | 0x80);
            value >>>= 7;
        }
    }
    writeZigZagInt(value: number): void {
        this.writeVarInt((value << 1) ^ (value >> 31));
    }
    static calcUtf8bytes(value: string): number {
        let bytes = 0;
        for (let c = 0; c < value.length; c++) {
            const code = value.charCodeAt(c);
            if (code < 0x80) {
                bytes += 1;
            } else if (code < 0x800) {
                bytes += 2;
            } else {
                bytes += 3;
            }
        }
        return bytes;
    }
    writeUtf8(value: string): void {
        for (let c = 0; c < value.length; c++) {
            const code = value.charCodeAt(c);
            if (code < 0x80) { // 0xxx xxxx
                this.writeByte(code);
            } else if (code < 0x800) { // 110x xxxx  10xx xxxx
                this.writeByte(0xC0 | ((code >> 6) & 0x1F));
                this.writeByte(0x80 | (code & 0x3F));
            } else { // 1110 xxxx  10xx xxxx  10xx xxxx
                this.writeByte(0xE0 | ((code >> 12) & 0x0F));
                this.writeByte(0x80 | ((code >> 6) & 0x3F));
                this.writeByte(0x80 | (code & 0x3F));
            }
        }
    }
    getArray(): Uint8Array {
        return this.array.subarray(0, this.position);
    }
}

export class Reader {
    readonly array: Uint8Array;
    private readonly length: number;
    private position = 0;
    constructor(arrayBuffer: ArrayBuffer) {
        this.array = new Uint8Array(arrayBuffer);
        this.length = arrayBuffer.byteLength;
    }
    needed(value: number): number {
        const oldPosition = this.position;
        this.position += value;
        if (this.position > this.length) {
            throw new Error("reader buffer underflow");
        }
        return oldPosition;
    }
    isEmpty(): boolean {
        return this.position >= this.length;
    }
    readByte(): number {
        return this.array[this.needed(1)];
    }
    readInt(): number {
        return new DataView(this.array.buffer).getInt32(this.needed(4));
    }
    readVarInt(): number {
        let shift = 0;
        let value = 0;
        while (shift < 32) {
            const b = this.readByte();
            value |= (b & 0x7F) << shift;
            if ((b & 0x80) === 0) {
                return value;
            }
            shift += 7;
        }
        throw new Error("malformed VarInt input");
    }
    readZigZagInt(): number {
        const value = this.readVarInt();
        return (value >>> 1) ^ -(value & 1);
    }
    readUtf8(bytes: number): string {
        let result = "";
        while (bytes-- > 0) {
            let code: number;
            const b1 = this.readByte();
            if ((b1 & 0x80) === 0) { // 0xxx xxxx
                code = b1;
            } else if ((b1 & 0xE0) === 0xC0) { // 110x xxxx  10xx xxxx
                const b2 = this.readByte();
                if ((b2 & 0xC0) !== 0x80) {
                    throw new Error("malformed String input (1)");
                }
                code = ((b1 & 0x1F) << 6) | (b2 & 0x3F);
                bytes--;
            } else if ((b1 & 0xF0) === 0xE0) { // 1110 xxxx  10xx xxxx  10xx xxxx
                const b2 = this.readByte();
                const b3 = this.readByte();
                if (((b2 & 0xC0) !== 0x80) || ((b3 & 0xC0) !== 0x80)) {
                    throw new Error("malformed String input (2)");
                }
                code = ((b1 & 0x0F) << 12) | ((b2 & 0x3F) << 6) | (b3 & 0x3F);
                bytes -= 2;
            } else {
                throw new Error("malformed String input (3)");
            }
            result += String.fromCharCode(code);
        }
        return result;
    }
}

export interface Serializer {
    read(reader: Reader): any;
    write(value: any, writer: Writer): void;
}

export function writeTo(serializer: Serializer, value: any): Uint8Array {
    const writer = new Writer(128);
    serializer.write(value, writer);
    return writer.getArray();
}

export function readFrom(serializer: Serializer, arrayBuffer: ArrayBuffer): any {
    const reader = new Reader(arrayBuffer);
    const value = serializer.read(reader);
    if (!reader.isEmpty()) {
        throw new Error("reader is not empty");
    }
    return value;
}

export interface TypeHandler<T> {
    read(reader: Reader, id2typeHandler?: TypeHandler<any>[]): T;
    write(value: T, writer: Writer): void;
}

export class TypeDesc {
    constructor(public readonly id: number, public readonly handler: TypeHandler<any>) {
        // empty
    }
    write(value: any, writer: Writer): void {
        writer.writeVarInt(this.id);
        this.handler.write(value, writer);
    }
}

class NullTypeHandler implements TypeHandler<any> {
    read(reader: Reader): any {
        return null;
    }
    write(value: any, writer: Writer): void {
        // empty
    }
}
const NULL_DESC = new TypeDesc(0, new NullTypeHandler);

class ListTypeHandler implements TypeHandler<any[]> {
    read(reader: Reader, id2typeHandler: TypeHandler<any>[]): any[] {
        const list: any[] = [];
        for (let size = reader.readVarInt(); size > 0; size--) {
            list.push(read(reader, id2typeHandler));
        }
        return list;
    }
    write(value: any[], writer: Writer): void {
        writer.writeVarInt(value.length);
        value.forEach(element => write(element, writer));
    }
}
export const LIST_DESC = new TypeDesc(2, new ListTypeHandler);

class BooleanTypeHandler implements TypeHandler<boolean> {
    read(reader: Reader): boolean {
        return reader.readByte() !== 0;
    }
    write(value: boolean, writer: Writer): void {
        writer.writeByte(value ? 1 : 0);
    }
}
export const BOOLEAN_DESC = new TypeDesc(3, new BooleanTypeHandler);

class NumberTypeHandler implements TypeHandler<number> {
    read(reader: Reader): number {
        return new DataView(reader.array.buffer).getFloat64(reader.needed(8));
    }
    write(value: number, writer: Writer): void {
        const position = writer.needed(8);
        new DataView(writer.array.buffer).setFloat64(position, value);
    }
}
export const NUMBER_DESC = new TypeDesc(4, new NumberTypeHandler);

class StringTypeHandler implements TypeHandler<string> {
    read(reader: Reader): string {
        return reader.readUtf8(reader.readVarInt());
    }
    write(value: string, writer: Writer): void {
        writer.writeVarInt(Writer.calcUtf8bytes(value));
        writer.writeUtf8(value);
    }
}
export const STRING_DESC = new TypeDesc(5, new StringTypeHandler);

class BytesTypeHandler implements TypeHandler<Uint8Array> {
    read(reader: Reader): Uint8Array {
        const length = reader.readVarInt();
        return new Uint8Array(reader.array.buffer, reader.needed(length), length);
    }
    write(value: Uint8Array, writer: Writer): void {
        writer.writeVarInt(value.length);
        const position = writer.needed(value.length);
        writer.array.set(value, position);
    }
}
export const BYTES_DESC = new TypeDesc(6, new BytesTypeHandler);

export const FIRST_DESC_ID = 7;

function read(reader: Reader, id2typeHandler: TypeHandler<any>[]): any {
    return id2typeHandler[reader.readVarInt()].read(reader, id2typeHandler);
}

function write(value: any, writer: Writer): void {
    if ((value === null) || (value === undefined)) {
        NULL_DESC.write(null, writer);
    } else if (typeof(value) === "number") {
        NUMBER_DESC.write(value, writer);
    } else if (typeof(value) === "string") {
        STRING_DESC.write(value, writer);
    } else if (typeof(value) === "boolean") {
        BOOLEAN_DESC.write(value, writer);
    } else if (Array.isArray(value)) {
        LIST_DESC.write(value, writer);
    } else if (value instanceof Uint8Array) {
        BYTES_DESC.write(value, writer);
    } else {
        const typeDesc = value.constructor.TYPE_DESC;
        if (typeDesc) {
            typeDesc.write(value, writer);
        } else {
            throw new Error("unexpected value '" + value + "'");
        }
    }
}

export abstract class Enum {
    protected constructor(public readonly number: number, public readonly name: string) {
        // empty
    }
    toString(): string {
        return this.name;
    }
}

class EnumTypeHandler implements TypeHandler<Enum> {
    constructor(private readonly values: Enum[]) {
        // empty
    }
    read(reader: Reader): Enum {
        return this.values[reader.readVarInt()];
    }
    write(value: Enum, writer: Writer): void {
        writer.writeVarInt(value.number);
    }
}

class FieldHandler {
    constructor(private readonly field: string, private readonly typeHandler: TypeHandler<any> | null) {
        // empty
    }
    read(object: any, reader: Reader, id2typeHandler: TypeHandler<any>[]): any {
        object[this.field] = this.typeHandler ? this.typeHandler.read(reader, id2typeHandler) : read(reader, id2typeHandler);
    }
    write(id: number, object: any, writer: Writer) {
        const value = object[this.field];
        if ((value !== null) && (value !== undefined)) {
            writer.writeVarInt(id);
            if (this.typeHandler) {
                this.typeHandler.write(value, writer);
            } else {
                write(value, writer);
            }
        }
    }
}

class ClassTypeHandler implements TypeHandler<any> {
    private readonly fieldId2handler: FieldHandler[] = [];
    constructor(private readonly Type: any) {
        // empty
    }
    addField(id: number, handler: FieldHandler): void {
        this.fieldId2handler[id] = handler;
    }
    read(reader: Reader, id2typeHandler: TypeHandler<any>[]): any {
        const object = new this.Type;
        while (true) {
            const id = reader.readVarInt();
            if (id === 0) {
                return object;
            }
            this.fieldId2handler[id].read(object, reader, id2typeHandler);
        }
    }
    write(value: any, writer: Writer): void {
        this.fieldId2handler.forEach((handler, id) => handler.write(id, value, writer));
        writer.writeVarInt(0);
    }
}

export class FastSerializer implements Serializer {
    private readonly id2typeHandler: TypeHandler<any>[] = [];
    constructor(...typeDescHolders: any[]) {
        const add = (typeDesc: TypeDesc) => {
            if (this.id2typeHandler[typeDesc.id]) {
                throw new Error("TypeDesc with id " + typeDesc.id + " already added");
            }
            this.id2typeHandler[typeDesc.id] = typeDesc.handler;
        };
        [NULL_DESC, LIST_DESC, BOOLEAN_DESC, NUMBER_DESC, STRING_DESC, BYTES_DESC].forEach(add);
        typeDescHolders.forEach(typeDescHolder => add(typeDescHolder.TYPE_DESC));
    }
    read(reader: Reader): any {
        return read(reader, this.id2typeHandler);
    }
    write(value: any, writer: Writer): void {
        write(value, writer);
    }
}

export class FieldDesc {
    constructor(public readonly id: number, public readonly name: string, public readonly typeDesc: TypeDesc | null) {
        // empty
    }
}

export function classDesc(id: number, Type: any, ...fieldDescs: FieldDesc[]): TypeDesc {
    const handler = new ClassTypeHandler(Type);
    fieldDescs.forEach(fieldDesc => {
        const typeDesc = fieldDesc.typeDesc;
        handler.addField(fieldDesc.id, new FieldHandler(fieldDesc.name, typeDesc && typeDesc.handler));
    });
    return new TypeDesc(id, handler);
}

export function enumValues(Type: any): Enum[] {
    const values: Enum[] = [];
    Object.keys(Type).forEach(property => {
        const value = Type[property];
        if (value instanceof Enum) {
            values[value.number] = value;
        }
    });
    return values;
}

export function enumDesc(id: number, Type: any): TypeDesc {
    return new TypeDesc(id, new EnumTypeHandler(Type.VALUES));
}

export abstract class Message {
    // empty
}

export class Request extends Message {
    constructor(public readonly serviceId: number, public readonly methodId: number, public readonly parameters: any[]) {
        super();
    }
}

export abstract class Reply extends Message {
    abstract process(): any;
}

export class ValueReply extends Reply {
    constructor(public readonly value: any) {
        super();
    }
    process(): any {
        return this.value;
    }
}

export class ExceptionReply extends Reply {
    constructor(public readonly exception: any) {
        super();
    }
    process(): any {
        throw this.exception;
    }
}

export class MessageSerializer implements Serializer {
    private static readonly REQUEST = 0;
    private static readonly VALUE_REPLY = 1;
    private static readonly EXCEPTION_REPLY = 2;
    constructor(private readonly contractSerializer: Serializer) {
        // empty
    }
    read(reader: Reader): Message {
        const type = reader.readByte();
        if (type === MessageSerializer.REQUEST) {
            return new Request(
                reader.readZigZagInt(),
                reader.readZigZagInt(),
                this.contractSerializer.read(reader)
            );
        } else if (type === MessageSerializer.VALUE_REPLY) {
            return new ValueReply(
                this.contractSerializer.read(reader)
            );
        } else {
            return new ExceptionReply(
                this.contractSerializer.read(reader)
            );
        }
    }
    write(message: Message, writer: Writer): void {
        if (message instanceof Request) {
            writer.writeByte(MessageSerializer.REQUEST);
            writer.writeZigZagInt(message.serviceId);
            writer.writeZigZagInt(message.methodId);
            this.contractSerializer.write(message.parameters, writer);
        } else if (message instanceof ValueReply) {
            writer.writeByte(MessageSerializer.VALUE_REPLY);
            this.contractSerializer.write(message.value, writer);
        } else {
            writer.writeByte(MessageSerializer.EXCEPTION_REPLY);
            this.contractSerializer.write((<ExceptionReply>message).exception, writer);
        }
    }
}

export class MethodMapping {
    constructor(public readonly method: string, public readonly id: number, public readonly oneWay: boolean) {
        // empty
    }
}

export class MethodMapper {
    private readonly id2mapping: MethodMapping[] = [];
    private readonly name2Mapping: {[methodName: string]: MethodMapping} = {};
    constructor(...mappings: MethodMapping[]) {
        mappings.forEach(mapping => {
            this.id2mapping[mapping.id] = mapping;
            this.name2Mapping[mapping.method] = mapping;
        });
    }
    mapId(id: number): MethodMapping {
        return this.id2mapping[id];
    }
    mapMethod(method: string): MethodMapping {
        return this.name2Mapping[method];
    }
    proxy(interceptor: (method: string, parameters: any[]) => any): any {
        const stub: any = {};
        Object.keys(this.name2Mapping).forEach(method => stub[method] = (...parameters: any[]) => interceptor(method, parameters));
        return stub;
    }
}

export class SimpleInterceptorContext {
    private static ID = 0;
    readonly id = SimpleInterceptorContext.ID++;
    constructor(public readonly methodMapping: MethodMapping, public readonly parameters: any[]) {
        // empty
    }
}

export interface Interceptor<C> {
    /**
     * @return context
     */
    entry(methodMapping: MethodMapping, parameters: any[]): C;
    exit(context: C, result: any): void;
    exception(context: C, exception: any): void;
    /**
     * Called after promise has been resolved.
     */
    resolved(context: C): void;
}

export const EMPTY_INTERCEPTOR: Interceptor<any> = {
    entry(methodMapping: MethodMapping, parameters: any[]): any {
        return 0;
    },
    exit(context: any, result: any): void {
        // empty
    },
    exception(context: any, exception: any): void {
        // empty
    },
    resolved(context: any): void {
        // empty
    }
};

export class ContractId<P, I> {
    constructor(public readonly id: number, public readonly methodMapper: MethodMapper) {
        // empty
    }
    service(implementation: I, interceptor = EMPTY_INTERCEPTOR): Service {
        return new Service(this, implementation, interceptor);
    }
    neededToPreventUnusedWarningOfTypeP(p: P): void {
        // empty
    }
}

export class Service {
    constructor(public readonly contractId: ContractId<any, any>, public readonly implementation: any, public readonly interceptor: Interceptor<any>) {
        // empty
    }
}

export abstract class AbstractInvocation {
    private context: any;
    protected constructor(public readonly methodMapping: MethodMapping, public readonly parameters: any[], private readonly interceptor: Interceptor<any>) {
        // empty
    }
    entry(): void {
        this.context = this.interceptor.entry(this.methodMapping, this.parameters);
    }
    exit(result: any): void {
        this.interceptor.exit(this.context, result);
    }
    exception(exception: any): void {
        this.interceptor.exception(this.context, exception);
    }
    resolved(): void {
        this.interceptor.resolved(this.context);
    }
}

export class ServerInvocation extends AbstractInvocation {
    constructor(public readonly service: Service, request: Request) {
        super(service.contractId.methodMapper.mapId(request.methodId), request.parameters, service.interceptor);
    }
    invoke(): Reply {
        this.entry();
        try {
            const implementation = this.service.implementation;
            let result = implementation[this.methodMapping.method].apply(implementation, this.parameters);
            result = ((result !== null) && (result !== undefined)) ? result : null;
            this.exit(result);
            return new ValueReply(result);
        } catch (exception) {
            this.exception(exception);
            return new ExceptionReply(exception);
        }
    }
}

export class Server {
    private readonly id2service: Service[] = [];
    constructor(...services: Service[]) {
        services.forEach(service => {
            const id = service.contractId.id;
            if (this.id2service[id]) {
                throw new Error("serviceId " + id + " already added");
            }
            this.id2service[id] = service;
        });
    }
    invocation(request: Request): ServerInvocation {
        const service = this.id2service[request.serviceId];
        if (!service) {
            throw new Error("no serviceId " + request.serviceId + " found (methodId " + request.methodId + ")");
        }
        return new ServerInvocation(service, request);
    }
    static readonly EMPTY = new Server;
}

export interface Tunnel {
    (request: Request): void;
}

export class ClientInvocation extends AbstractInvocation {
    readonly promise: Promise<any>;
    settle: (reply: Reply) => void;
    constructor(methodMapping: MethodMapping, parameters: any[], interceptor: Interceptor<any>, private readonly serviceId: number) {
        super(methodMapping, parameters, interceptor);
        if (methodMapping.oneWay) {
            return;
        }
        this.promise = new Promise<any>((resolve, reject) => {
            this.settle = reply => {
                try {
                    this.exit(reply.process());
                } catch (exception) {
                    this.exception(exception);
                }
                try {
                    resolve(reply.process());
                } catch (exception) {
                    reject(exception);
                }
                Promise.resolve().then(() => {
                    this.resolved();
                });
            };
        });
    }
    invoke(tunnel: Tunnel): void {
        this.entry();
        tunnel(new Request(this.serviceId, this.methodMapping.id, this.parameters));
    }
}

export abstract class Client {
    proxy<P>(contractId: ContractId<P, any>, interceptor = EMPTY_INTERCEPTOR): P {
        return contractId.methodMapper.proxy((method, parameters) => {
            const invocation = new ClientInvocation(contractId.methodMapper.mapMethod(method), parameters, interceptor, contractId.id);
            this.invoke(invocation);
            return invocation.promise;
        });
    }
    protected abstract invoke(invocation: ClientInvocation): void;
}

export class Packet {
    static readonly END_REQUESTNUMBER = 0;
    static readonly END = new Packet(Packet.END_REQUESTNUMBER, null);
    constructor(public readonly requestNumber: number, public readonly message: Message | null) {
        // empty
    }
    isEnd(): boolean {
        return this.requestNumber === Packet.END_REQUESTNUMBER;
    }
}

export class PacketSerializer implements Serializer {
    constructor(private readonly messageSerializer: Serializer) {
        // empty
    }
    read(reader: Reader): Packet {
        const requestNumber = reader.readInt();
        return (requestNumber === Packet.END_REQUESTNUMBER) ? Packet.END : new Packet(requestNumber, this.messageSerializer.read(reader));
    }
    write(packet: Packet, writer: Writer): void {
        if (packet.isEnd()) {
            writer.writeInt(Packet.END_REQUESTNUMBER);
        } else {
            writer.writeInt(packet.requestNumber);
            this.messageSerializer.write(packet.message, writer);
        }
    }
}

export interface Connection {
    write(packet: Packet): void;
    closed(): void;
}

export interface SessionFactory {
    /**
     * It's allowed to call Client.proxy(ContractId, Interceptor...) during this method,
     * but the proxies can be used not before Session.opened() is called.
     * If this method throws an exception, the connection is rejected and Session.closed(Exception) won't be called.
     */
    (connection: Connection): Session;
}

export abstract class Session extends Client {
    protected constructor(private readonly connection: Connection) {
        super();
    }
    private serverProp: Server;
    /**
     * Gets the server of this session. Called only once after creation of session.
     * This implementation returns Server.EMPTY.
     */
    protected server(): Server {
        return Server.EMPTY;
    }
    /**
     * Called when the session has been opened.
     * This implementation does nothing.
     */
    protected opened(): void {
        // empty
    }
    /**
     * Called once when the session has been closed.
     * This implementation does nothing.
     * @param exception if (exception == null) regular close else reason for close
     */
    protected closed(exception: any): void {
        // empty
    }
    private closedProp = true;
    isClosed(): boolean {
        return this.closedProp;
    }
    private doCloseSend(sendEnd: boolean, exception: any): void {
        if (this.closedProp) {
            return;
        }
        this.closedProp = true;
        try {
            this.closed(exception);
            if (sendEnd) {
                this.connection.write(Packet.END);
            }
        } finally {
            this.connection.closed();
        }
    }
    static doClose(session: Session, exception: any): void {
        session.doCloseSend(false, exception);
    }
    /**
     * Closes the session.
     * This method is idempotent.
     */
    close(): void {
        this.doCloseSend(true, null);
    }
    private serverInvoke(requestNumber: number, request: Request): void {
        const invocation = this.serverProp.invocation(request);
        const reply = invocation.invoke();
        if (!invocation.methodMapping.oneWay) {
            this.connection.write(new Packet(requestNumber, reply));
        }
    }
    private requestNumber2invocation: ClientInvocation[] = [];
    private received(packet: Packet): void {
        try {
            if (packet.isEnd()) {
                this.doCloseSend(false, null);
                return;
            }
            const message = packet.message;
            if (message instanceof Request) {
                this.serverInvoke(packet.requestNumber, message);
            } else { // client invoke
                const invocation = this.requestNumber2invocation[packet.requestNumber];
                delete this.requestNumber2invocation[packet.requestNumber];
                invocation.settle(<Reply>message);
            }
        } catch (exception) {
            Session.doClose(this, exception);
            throw exception;
        }
    }
    static received(session: Session, packet: Packet) {
        session.received(packet);
    }
    private requestNumber = Packet.END_REQUESTNUMBER;
    protected invoke(invocation: ClientInvocation): void {
        if (this.isClosed()) {
            throw new Error("session is already closed or not yet opened");
        }
        invocation.invoke(request => {
            try {
                if (this.requestNumber === 2147483647) {
                    this.requestNumber = Packet.END_REQUESTNUMBER;
                }
                this.requestNumber++;
                if (!invocation.methodMapping.oneWay) {
                    this.requestNumber2invocation[this.requestNumber] = invocation;
                }
                this.connection.write(new Packet(this.requestNumber, request));
            } catch (exception) {
                Session.doClose(this, exception);
                throw exception;
            }
        });
    }
    private created(): void {
        this.closedProp = false;
        try {
            this.serverProp = this.server();
            this.opened();
        } catch (e) {
            Session.doClose(this, e);
        }
    }
    static create(sessionFactory: SessionFactory, connection: Connection): Session {
        const session = sessionFactory(connection);
        session.created();
        return session;
    }
}

export function connect(url: string, serializer: Serializer, sessionFactory: SessionFactory): void {
    serializer = new PacketSerializer(new MessageSerializer(serializer));
    const ws = new WebSocket(url);
    ws.binaryType = "arraybuffer";
    ws.onerror = () => {
        throw new Error("WebSocket.onerror");
    };
    ws.onclose = () => {
        throw new Error("WebSocket.onclose");
    };
    ws.onopen = () => {
        try {
            const session = Session.create(sessionFactory, {
                write: packet => ws.send(writeTo(serializer, packet)),
                closed: () => ws.close()
            });
            ws.onmessage = event => {
                try {
                    Session.received(session, readFrom(serializer, event.data));
                } catch (e) {
                    Session.doClose(session, e);
                }
            };
            ws.onerror = () => Session.doClose(session, new Error("WebSocket.onerror"));
            ws.onclose = event => {
                if (event.wasClean) {
                    session.close();
                } else {
                    Session.doClose(session, new Error("WebSocket.onclose"));
                }
            }
        } catch (exception) {
            ws.close();
            throw exception;
        }
    };
}

class XhrClient extends Client {
    private readonly serializer: Serializer;
    constructor(private readonly url: string, serializer: Serializer, private readonly timeoutMilliSeconds: number) {
        super();
        this.serializer = new MessageSerializer(serializer);
    }
    protected invoke(invocation: ClientInvocation): void {
        invocation.invoke(request => {
            const xhr = new XMLHttpRequest();
            xhr.open("POST", this.url);
            xhr.responseType = "arraybuffer";
            xhr.timeout = this.timeoutMilliSeconds;
            xhr.ontimeout = () => invocation.settle(new ExceptionReply(new Error("XMLHttpRequest.ontimeout")));
            xhr.onerror = () => invocation.settle(new ExceptionReply(new Error("XMLHttpRequest.onerror")));
            xhr.onload = () => {
                try {
                    invocation.settle(readFrom(this.serializer, xhr.response));
                } catch (exception) {
                    invocation.settle(new ExceptionReply(exception));
                }
            };
            xhr.send(writeTo(this.serializer, request));
        });
    }
}

export function xhr(url: string, serializer: Serializer, timeoutMilliSeconds = 0): Client {
    return new XhrClient(url, serializer, timeoutMilliSeconds);
}
